python/thunder/utils/decorators.py

```diff
@@ -0,0 +1,281 @@
+""" Useful decorators that are used throughout the library """
+
+def _isNamedTuple(obj):
+  """Heuristic check if an object is a namedtuple."""
+  from collections import namedtuple
+  return hasattr(obj, "_fields") \
+     and hasattr(obj, "_asdict") \
+     and callable(obj._asdict)
+
+def serializable(cls):
+    '''The @serializable decorator can decorate any class to make it easy to store
+    that class in a human readable JSON format and then recall it and recover
+    the original object instance. Classes instances that are wrapped in this
+    decorator gain the serialize() method, and the class also gains a
+    deserialize() static method that can automatically "pickle" and "unpickle" a
+    wide variety of objects like so:
+
+      @serializable
+      class Visitor():
+          def __init__(self, ipAddr = None, agent = None, referrer = None):
+              self.ip = ipAddr
+              self.ua = agent
+              self.referrer= referrer
+              self.time = datetime.datetime.now()
+
+      origVisitor = Visitor('192.168', 'UA-1', 'http://www.google.com')
+
+      #serialize the object
+      pickledVisitor = origVisitor.serialize()
+
+      #restore object
+      recovVisitor = Visitor.deserialize(pickledVisitor)
+
+    Note that this decorator is NOT designed to provide generalized pickling
+    capabilities. Rather, it is designed to make it very easy to convert small
+    classes containing model properties to a human and machine parsable format
+    for later analysis or visualization. A few classes under consideration for
```

Comment:
Maybe drop this last sentence, great for our own planning, but will (hopefully!) be only a transient a description of the state of things and thus not really for the docs.